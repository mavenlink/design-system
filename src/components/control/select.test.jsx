import React, { createRef } from 'react';
import {
  render as _render,
  screen,
} from '@testing-library/react';
import user from '@testing-library/user-event';
import Select from './select.jsx';

const render = (ui, options = { labelledBy: 'labelled-by' }) => (
  _render(ui, {
    ...options,
    wrapper: props => (
      <>
        <label htmlFor="test-id" id="test-label-id">Test label</label>
        {props.children}
      </>
    ),
  })
);
describe('src/components/control/select', () => {
  const requiredProps = {
    id: 'test-id',
    labelledBy: 'test-label-id',
    listOptionRefs: [],
    name: 'field-id',
    wrapperRef: null,
  };

  it('has defaults', () => {
    const ref = createRef();
    render(<Select {...requiredProps} ref={ref} />);
    expect(document.body).toMatchSnapshot();
    expect(ref.current).toMatchSnapshot();
    expect(document.activeElement).toBe(document.body);
    user.click(screen.getByLabelText('Test label'));
    expect(document.body).toMatchSnapshot();
  });

  it('displays (loading) empty objects', () => {
    const { rerender } = render(<Select {...requiredProps} value={{ id: 8 }} displayValueEvaluator={o => o.label} />);
    expect(screen.getByLabelText('Test label')).toHaveValue('');
    rerender(<Select {...requiredProps} value={{ id: 8, label: 'foo' }} displayValueEvaluator={o => o.label} />);
    expect(screen.getByLabelText('Test label')).toHaveValue('foo');
  });

  it('resets the searchValue when value is set to undefined (regression test)', () => {
    const { rerender } = render(<Select {...requiredProps} value={{ id: 8, label: 'foo' }} displayValueEvaluator={o => o.label} />);
    expect(screen.getByLabelText('Test label')).toHaveValue('foo');
    rerender(<Select {...requiredProps} value={undefined} displayValueEvaluator={o => o.label} />);
    expect(screen.getByLabelText('Test label')).toHaveValue('');
  });
});
