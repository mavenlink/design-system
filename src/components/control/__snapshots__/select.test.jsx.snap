// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/control/select classNames API sets all class names 1`] = `
<body>
  <div>
    <label
      for="test-id"
      id="test-label-id"
    >
      Test label
    </label>
    <div
      class="unique-container"
      style="position: relative;"
    >
      <input
        aria-autocomplete="none"
        aria-controls="test-id-single-choice-listbox"
        aria-describedby="test-id-tooltip test-idHint"
        aria-expanded="false"
        aria-haspopup="listbox"
        autocomplete="off"
        class="unique-input"
        id="test-id"
        name="field-id"
        role="combobox"
        style="--numIcon: 3;"
        type="text"
        value=""
      />
      <div
        class="container"
      >
        <svg
          aria-disabled="false"
          class="button"
          height="6"
          role="button"
          tabindex="0"
          width="10"
        >
          <title>
            Open choices listbox
          </title>
          <use
            xlink:href="#caret-down.svg"
          />
        </svg>
      </div>
    </div>
  </div>
</body>
`;

exports[`src/components/control/select classNames API sets container 1`] = `
<body>
  <div>
    <label
      for="test-id"
      id="test-label-id"
    >
      Test label
    </label>
    <div
      class="unique-container"
      style="position: relative;"
    >
      <input
        aria-autocomplete="none"
        aria-controls="test-id-single-choice-listbox"
        aria-describedby="test-id-tooltip test-idHint"
        aria-expanded="false"
        aria-haspopup="listbox"
        autocomplete="off"
        class="input"
        id="test-id"
        name="field-id"
        role="combobox"
        style="--numIcon: 3;"
        type="text"
        value=""
      />
      <div
        class="container"
      >
        <svg
          aria-disabled="false"
          class="button"
          height="6"
          role="button"
          tabindex="0"
          width="10"
        >
          <title>
            Open choices listbox
          </title>
          <use
            xlink:href="#caret-down.svg"
          />
        </svg>
      </div>
    </div>
  </div>
</body>
`;

exports[`src/components/control/select classNames API sets valid input 1`] = `
<body>
  <div>
    <label
      for="test-id"
      id="test-label-id"
    >
      Test label
    </label>
    <div
      class="container"
      style="position: relative;"
    >
      <input
        aria-autocomplete="none"
        aria-controls="test-id-single-choice-listbox"
        aria-describedby="test-id-tooltip test-idHint"
        aria-expanded="false"
        aria-haspopup="listbox"
        autocomplete="off"
        class="unique-input"
        id="test-id"
        name="field-id"
        role="combobox"
        style="--numIcon: 3;"
        type="text"
        value=""
      />
      <div
        class="container"
      >
        <svg
          aria-disabled="false"
          class="button"
          height="6"
          role="button"
          tabindex="0"
          width="10"
        >
          <title>
            Open choices listbox
          </title>
          <use
            xlink:href="#caret-down.svg"
          />
        </svg>
      </div>
    </div>
  </div>
</body>
`;

exports[`src/components/control/select has defaults 1`] = `
<body>
  <div>
    <label
      for="test-id"
      id="test-label-id"
    >
      Test label
    </label>
    <div
      class="container"
      style="position: relative;"
    >
      <input
        aria-autocomplete="none"
        aria-controls="test-id-single-choice-listbox"
        aria-describedby="test-id-tooltip test-idHint"
        aria-expanded="false"
        aria-haspopup="listbox"
        autocomplete="off"
        class="input"
        id="test-id"
        name="field-id"
        role="combobox"
        style="--numIcon: 3;"
        type="text"
        value=""
      />
      <div
        class="container"
      >
        <svg
          aria-disabled="false"
          class="button"
          height="6"
          role="button"
          tabindex="0"
          width="10"
        >
          <title>
            Open choices listbox
          </title>
          <use
            xlink:href="#caret-down.svg"
          />
        </svg>
      </div>
    </div>
  </div>
</body>
`;

exports[`src/components/control/select has defaults 2`] = `
Object {
  "dirty": false,
  "name": "field-id",
  "value": undefined,
}
`;

exports[`src/components/control/select has defaults 3`] = `
<body>
  <div>
    <label
      for="test-id"
      id="test-label-id"
    >
      Test label
    </label>
    <div
      class="container"
      style="position: relative;"
    >
      <input
        aria-autocomplete="none"
        aria-controls="test-id-single-choice-listbox"
        aria-describedby="test-id-tooltip test-idHint"
        aria-expanded="true"
        aria-haspopup="listbox"
        autocomplete="off"
        class="input"
        id="test-id"
        name="field-id"
        role="combobox"
        style="--numIcon: 3;"
        type="text"
        value=""
      />
      <div
        class="container"
      >
        <svg
          aria-disabled="false"
          class="button"
          height="6"
          role="button"
          tabindex="0"
          width="10"
        >
          <title>
            Open choices listbox
          </title>
          <use
            xlink:href="#caret-down.svg"
          />
        </svg>
      </div>
      <ul
        aria-labelledby="test-id-label"
        class="dropdown"
        id="test-id-single-choice-listbox"
        role="listbox"
      >
        <span
          class="no-options"
        >
          No options available.
        </span>
      </ul>
    </div>
  </div>
</body>
`;
